TOML: initial formatter implementation

Kobzol opened the pull request:
This PR implements an initial formatter for TOML. I copied the Rust formatter and stripped it down to only keep things that were required to implement basic spacing formatting.

The [official repository](https://github.com/toml-lang/toml) does not mandate any formatting style and I think that it is a bit inconsistent (space after `[` in outer array, but not in nested array, while for objects it uses spaces after `{` even for nested objects).
There is https://github.com/jumpscale7/python-pretty-toml/blob/master/sample-prettified.toml, but it's also not very specified.

Right now there is only a very basic spacing implementation. Some things are probably OK to enforce (space after `,`, space before and after `=`), but for controversial topics like space after `[`, we should probably introduce an option in TOML code style settings.

After we resolve that, I think that this implementation is already quite useful and could be merged, but we should deal with indents somehow. Right now only `noneIndent` is used and so this:
```rust
multiline = [
      1,
      2
]
```
turns into
```rust
multiline = [
1,
2
]
```
which is not ideal, the explicit indent should be preserved.

Fixes: https://github.com/intellij-rust/intellij-rust/issues/5518

Undin commented on intellij-toml/src/main/kotlin/org/toml/ide/formatter/settings/TomlLanguageCodeStyleSettingsProvider.kt:
```diff
@@ -0,0 +1,46 @@
+        }
+    }
+
+    override fun getIndentOptionsEditor(): IndentOptionsEditor? = SmartIndentOptionsEditor()
```
If we decide not to provide indent settings in this PR (i.e. `org.toml.ide.formatter.settings.TomlCodeStyleSettingsProvider.TomlCodeStyleMainPanel#initTabs` initializes only space tab), it will be redudant

Undin commented on intellij-toml/src/test/kotlin/org/toml/ide/formatter/TomlFormatterTest.kt:
```diff
@@ -0,0 +1,10 @@
+package org.toml.ide.formatter
+
+class TomlFormatterTest : TomlFormatterTestBase() {
+    fun `test spacing`() = doTest()
```
Let's better inline test texts.
I find it much easier to have code example in test class directly instead of looking for before/after files in test data

Kobzol commented:
I inlined the formatter tests, added a typing test and a simple base class for `TOML` tests. I also added a simple indent implementation that handles array (and enter being pressed inside an array). I changed the settings to include the basic indent stuff like tabs/spaces and indent size.

Since objects (inline tables) have to be on a single line, I think that arrays are actually the only thing in TOML that require indent handling? In such case I think that the current implementation would be fine as a MVP.

In the TOML reference there are indented sub tables:
```toml
[servers]
  # Indentation (tabs and/or spaces) is allowed but not required
  [servers.alpha]
  ip = "10.0.0.1"
  dc = "eqdc10"

  [servers.beta]
  ip = "10.0.0.2"
  dc = "eqdc10"
```
I haven't seen this in the wild yet, but maybe we can include it as an option in the indent settings.

Should I still turn this as an experimental feature now that basic indenting works?

Undin commented on intellij-toml/src/test/kotlin/org/toml/TomlTestBase.kt:
```diff
@@ -0,0 +1,28 @@
+    }
+}
+
+fun replaceCaretMarker(text: String): String = text.replace("/*caret*/", "<caret>")
```
Actually, this function doesn't make sense for TOML. `/*caret*/` is valid block comment in Rust language and it was introduced to avoid syntax errors in injected code.
But TOML langugage doesn't have block comments at all so I don't know how to avoid syntax error when we need caret symbol in TOML code

Kobzol replied:
Thanks, that makes sense.

Undin commented on intellij-toml/src/test/kotlin/org/toml/TomlTestBase.kt:
```diff
@@ -0,0 +1,28 @@
+
+abstract class TomlTestBase: BasePlatformTestCase() {
+    @Suppress("TestFunctionName")
+    protected fun InlineFile(@Language("TOML") text: String, name: String = "Cargo.toml") {
```
```suggestion
    protected fun InlineFile(@Language("TOML") text: String, name: String = "example.toml") {
```

Just non to mislead anyone that these tests somehow related to Rust and Cargo

Undin commented on intellij-toml/src/test/kotlin/org/toml/TomlTestBase.kt:
```diff
@@ -0,0 +1,28 @@
+
+    protected fun checkByText(
+        @Language("Rust") before: String,
+        @Language("Rust") after: String,
```
`Rust` -> `TOML`

Undin commented on intellij-toml/src/test/kotlin/org/toml/ide/formatter/TomlAutoIndentTest.kt:
```diff
@@ -0,0 +1,23 @@
+
+class TomlAutoIndentTest : TomlTypingTestBase() {
+    fun `test new array element`() = doTestByText("""
+    [key]
+    foo = [
+        "text",/*caret*/
+    ]
+    """, """
+    [key]
+    foo = [
+        "text",
+        /*caret*/
+    ]
```
```suggestion
        [key]
        foo = [
            "text",/*caret*/
        ]
        """, """
        [key]
        foo = [
            "text",
            /*caret*/
        ]
```

Undin commented on intellij-toml/src/test/kotlin/org/toml/ide/formatter/TomlFormatterTestBase.kt:
```diff
@@ -0,0 +1,25 @@
+
+abstract class TomlFormatterTestBase : TomlTestBase() {
+    protected fun doTest(@Language("TOML") before: String, @Language("TOML") after: String) {
+        InlineFile(before.trim())
+        WriteCommandAction.runWriteCommandAction(project) {
+            val file = myFixture.file
+            CodeStyleManager.getInstance(project)
+                .reformatText(file, file.textRange.startOffset, file.textRange.endOffset)
+        }
+
+        myFixture.checkResult(after.trim())
+    }
```
Let's use `checkByText` here

Undin commented on intellij-toml/src/test/kotlin/org/toml/ide/formatter/TomlFormatterTest.kt:
```diff
@@ -0,0 +1,36 @@
+package org.toml.ide.formatter
+
+class TomlFormatterTest : TomlFormatterTestBase() {
+    fun `test spacing`() = doTest("""
+[  config . subconfig ]
+    list  =   [  1,2,     3   ]
+key1 . key2='value'
+object={a=1,b=2}
+    """, """
+[config.subconfig]
+list = [1, 2, 3]
+key1.key2 = 'value'
+object = { a = 1, b = 2 }
+    """)
```
```suggestion
    fun `test spacing`() = doTest("""
        [  config . subconfig ]
            list  =   [  1,2,     3   ]
        key1 . key2='value'
        object={a=1,b=2}
    """, """
        [config.subconfig]
        list = [1, 2, 3]
        key1.key2 = 'value'
        object = { a = 1, b = 2 }
    """)
```

I know that test will not pass without additional changes. Suggest using `trimIndent` for expected/actual texts in some base method. For example, in `checkByText`

Undin commented on intellij-toml/src/test/kotlin/org/toml/ide/typing/RsTypingTestBase.kt:
```diff
@@ -0,0 +1,16 @@
+import org.intellij.lang.annotations.Language
+import org.toml.TomlTestBase
+
+abstract class TomlTypingTestBase : TomlTestBase() {
```
Let's change file name to `TomlTypingTestBase`

Undin commented on intellij-toml/src/test/kotlin/org/toml/TomlTestBase.kt:
```diff
@@ -0,0 +1,28 @@
+import com.intellij.testFramework.fixtures.BasePlatformTestCase
+import org.intellij.lang.annotations.Language
+
+abstract class TomlTestBase: BasePlatformTestCase() {
```
nit: formatting

Kobzol commented:
Ouch, nasty copy paste errors. Thanks.

Undin commented on intellij-toml/src/test/kotlin/org/toml/ide/formatter/TomlFormatterTest.kt:
```diff
@@ -0,0 +1,36 @@
+            list  =   [  1,2,     3   ]
+        key1 . key2='value'
+        object={a=1,b=2}
+            """, """
```
```suggestion
    """, """
```

Undin commented on intellij-toml/src/test/kotlin/org/toml/ide/formatter/TomlFormatterTest.kt:
```diff
@@ -0,0 +1,36 @@
+                "b.rs",
+                    "c.rs"
+        ]
+            """, """
```
The same here

Undin commented on intellij-toml/src/main/kotlin/org/toml/ide/formatter/settings/TomlCodeStyleSettingsProvider.kt:
```diff
@@ -0,0 +1,32 @@
+class TomlCodeStyleSettingsProvider : CodeStyleSettingsProvider() {
+    override fun createCustomSettings(settings: CodeStyleSettings) = TomlCodeStyleSettings(settings)
+
+    override fun getConfigurableDisplayName() = TomlLanguage.displayName
```
nit: Could you specify return type explicitly?

Kobzol commented:
bors r+
